doc/design/proposal-moving-hooks-on-thread.md

```diff
@@ -0,0 +1,160 @@
+
+Because this hook might need to resolve and load the entire module graph to do its work, such as to resolve and fetch HTTPS URLs, it will be provided with the `resolve` and `load` hooks in its `context`. It would also receive `context.entry` to distinguish main/worker entry points from dynamic `import()`. Its signature would be something like:
+
+```js
+/**
+ * @typedef {{
+ *   parentURL?: string,
+ *   entry: boolean,
+ *   conditions?: string[],
+ *   importAttributes?: Record<string, string>,
+ *   resolve: Function,
+ *   load: Function,
+ * }} ModuleStartGraphContext
+ */
+/**
+ * @typedef {Promise<{
+ *   shortCircuit?: boolean
+ * }>} ModuleStartGraphResult
+ */
+/**
+ * @param {string} specifier
+ * @param {ModuleStartGraphContext} context
+ * @param {(specifier: string, context: ModuleStartGraphContext) => ModuleStartGraphResult} nextStartGraph
+ * @returns {ModuleStartGraphResult}
+ */
+async function startGraph(specifier, context, nextStartGraph) {}
+```
+
+## Examples
+
+### Import from HTTPS
+
+the [Import from HTTPS example](https://nodejs.org/api/module.html#import-from-https) in our docs could use `startGraph` to handle the async work:
+
+```js
+import { parseAllImportSpecifiersFromSource } from 'some-library';
+
+/** @type {Map<URL['href'], string | Buffer} */
+const httpImportsCache = new Map();
+
+export async function startGraph(specifier, context, nextStartGraph) {
+  const { resolve, load } = context;
+  const { url } = resolve(specifier, context);
+  const urls = [url];
+  while (urls.length > 0) {
+    const url = urls.shift();
+    if (url.startsWith('https://')) {
+      const response = await fetch(url);
+      const responseSource = await response.text();
+      httpImportsCache.set(url, responseSource);
+    }
+    // load API constructs context automatically from registry data
+    const { source } = load(url);
+    httpImportsCache.set(url, source);
+    for (const specifier of parseAllImportSpecifiersFromSource(source)) {
+      // resolve API also constructs context automatically from registry data
+      // including determining if the parent is CJS or ESM and setting conditions
+      // correctly. Even when conditions is provided, although explicit import
+      // or require conditions can override this process.
+      const depUrl = resolve(specifier, { parentURL: url });
+      urls.push(depUrl);
+    }
+  }
+  return nextStartGraph(specifier, context);
+}
+
+export function resolve(specifier, context, defaultResolve) {
+  if (specifier.startsWith('https://')) {
+    return {
+      url: specifier,
+      shortCircuit: true,
+    };
+  }
+  return defaultResolve(specifier, context);
+}
+
+export function load(url, context, defaultLoad) {
```

Comment:
Same question as with `resolve`. What about merely requiring `{ loaded: true || false }` as the returned result?

Reply:
This document doesn’t propose any changes to `load`. It’s only mentioned here as part of an example.